m2cgen/interpreters/mixins.py

```diff
@@ -0,0 +1,87 @@
+import sys
+
+from m2cgen import ast
+from m2cgen.interpreters.interpreter import BaseAstToCodeInterpreter
+
+
+class BinExpressionDepthTrackingMixin(BaseAstToCodeInterpreter):
+    """
+    This mixin provides an ability to call a custom cook when depth of the
+    binary expression reaches certain threshold.
+
+    Subclasses must specify value for `bin_depth_threshold`
+
+    By default it creates a variable and assigns it the result of the incoming
+    expression interpretation.
+
+    Subclasses may override this default behaviour.
+    """
+
+    # disabled by default
+    bin_depth_threshold = sys.maxsize
+
+    def _pre_interpret_hook(self, expr, bin_depth=0, **kwargs):
+        if not isinstance(expr, ast.BinExpr):
+            return None, kwargs
+
+        # We track depth of the binary expressions and call a hook if it
+        # reaches specified threshold .
+        if bin_depth == self.bin_depth_threshold:
+            return self.bin_depth_threshold_hook(expr, **kwargs), kwargs
+
+        kwargs["bin_depth"] = bin_depth + 1
+        return None, kwargs
+
+    # Default implementation. Simply adds new variable.
+    def bin_depth_threshold_hook(self, expr, **kwargs):
+        var_name = self._cg.add_var_declaration(expr.output_size)
+        result = self._do_interpret(expr, **kwargs)
+        self._cg.add_var_assignment(var_name, result, expr.output_size)
+        return var_name
+
+
+class LinearAlgebraMixin(BaseAstToCodeInterpreter):
+    """
+    This mixin provides simple way to interpret linear algebra expression as
+    function invocation.
+
+    It also provides flag `with_linear_algebra` which indicates whether
+    linear algebra was used during interpretation. It can be used to add
+    dependencies.
```

Comment:
Maybe `add corresponding third party dependencies that provide linear algebra operations and/or data structures`?

Reply:
Awesome, thanks!